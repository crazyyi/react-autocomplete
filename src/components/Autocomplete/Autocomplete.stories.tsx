import { Autocomplete } from "./Autocomplete";
import type { Meta, StoryObj } from '@storybook/react';
import { AutocompleteProps } from "./Autocomplete.types";
import { v4 } from "uuid";

const meta: Meta<typeof Autocomplete> = {
  title: "Components/Autocomplete",
  component: Autocomplete,
  argTypes: {

  }
}

export default meta

type Story = StoryObj<typeof Autocomplete>

const Template = (args: AutocompleteProps) => (
  <Autocomplete {...args} />
)

export const Default: Story = {
  render: Template,
  args: {
    label: "Auto Complete",
    renderItem: () => (value) => <li key={v4()} className="listitem">{value}</li>,
    list: [
      "white",
      "black",
      "yellow",
      "red",
      "gray",
      "blue",
      "orange",
      "purple",
      "green",
      "golden"
    ]
  }
}