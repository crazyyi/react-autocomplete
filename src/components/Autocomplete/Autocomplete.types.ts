import { ReactElement, Ref } from "react";
import { ItemProps } from "../List/Item";

interface Props<T> {
  /**
     * Ref to the DOM node.
     */
  ref?: Ref<HTMLElement | null>;
  label: string;
  list: string[];
  renderItem?: (itemProps: ItemProps) => (value: string) => ReactElement<ItemProps>
}

export type AutocompleteProps<T = object> = Props<T>