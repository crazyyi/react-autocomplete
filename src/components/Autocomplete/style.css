input {
  width: 120px;
}

.autocomplete-wrap {
  width: 120px;
}

ul {
  list-style-type: none;
  padding: 0px;
  margin: 0px;
  width: 100%;
  border: 1px solid black;
}

.listitem {
  padding: 4px;
  margin: 0px;
  background-color: #f0f0f0; /* Default background color */
  cursor: pointer;
}

.listitem:hover {
  background-color: #c0c0c0; 
}