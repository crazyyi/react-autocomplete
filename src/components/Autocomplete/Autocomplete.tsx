import { ComponentProps, useCallback, useRef } from "react"
import { AutocompleteProps } from "./Autocomplete.types"
import { useOutsideClick } from "../../utils/hooks"
import { v4 } from 'uuid'
import "./style.css"

export const Autocomplete = (
  {
    label,
    list,
    renderItem,
    ...props
  }: AutocompleteProps) => {
  const refDiv = useRef<HTMLDivElement>(null)
  const { show, clickHandler } = useOutsideClick(refDiv.current)

  const itemProps = {

  }

  const renderInput = ({ domProps }: { domProps?: ComponentProps<'input'> } = { domProps: {} }) => (
    <input data-testid="input" {...domProps} ></input>
  )

  const renderList = useCallback(({ list, domProps = {} }: {
    list: string[], domProps?: ComponentProps<'ul'>
  }) => {
    return list && list.length > 0 && <ul data-testid="listbox" aria-label="datalist" className="list" {...domProps}>{list.map((value) => <li key={v4()}>{renderItem?.({})?.(value)}</li>)}</ul>
  }, [renderItem])

  const renderError = () => (
    <div data-testid="error"></div>
  )

  return <div data-testid="wrapper" ref={refDiv} className="autocomplete-wrap" onClick={clickHandler}>
    {renderInput()}
    {show && renderList({ list })}
    {renderError()}
  </div>
}