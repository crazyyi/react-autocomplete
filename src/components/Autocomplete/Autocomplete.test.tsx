import { render, screen, within } from '@testing-library/react';
import { AutocompleteProps } from './Autocomplete.types';
import { Autocomplete } from './Autocomplete'; // Replace with actual path
import userEvent from "@testing-library/user-event";

jest.mock('./Autocomplete.types'); // Mock AutocompleteProps for now

describe('Autocomplete component', () => {
  const mockList = ['black', 'white', 'red', 'yellow'];
  const mockRenderItem = jest.fn().mockImplementation(() => (value: string) => <div>{value}</div>);
  let mockAutocompleteProps: AutocompleteProps;

  beforeEach(() => {
    mockAutocompleteProps = {
      label: "",
      list: mockList,
      renderItem: mockRenderItem,
    };
  })

  it('should match snapshot', () => {
    mockAutocompleteProps.label = "hello"
    const { container } = render(<Autocomplete {...mockAutocompleteProps} />);
    expect(container).toMatchSnapshot()
  })

  it('renders the input element', async () => {
    mockAutocompleteProps.label = "hello"
    render(<Autocomplete {...mockAutocompleteProps} />);
    expect(await screen.findByTestId('input')).toBeTruthy()
  });

  it('toggles the dropdown list on click', async () => {
    mockAutocompleteProps.label = "world"
    render(<Autocomplete {...mockAutocompleteProps} />);
    const wrapper = await screen.findByTestId('wrapper');
    await userEvent.click(wrapper)

    expect(await screen.findByTestId('listbox')).toBeTruthy();
  });

  it('hides the dropdown list when clicked outside', async () => {
    mockAutocompleteProps.label = "hello"
    render(<Autocomplete {...mockAutocompleteProps} />);
    const wrapper = await screen.findByTestId('wrapper');
    await userEvent.click(wrapper)

    expect(await screen.findByTestId('listbox')).toBeTruthy();

    await userEvent.click(document.body)

    expect(screen.queryByRole('listbox')).toBeNull();
  });

  it('renders list items using renderItem prop', async () => {
    mockAutocompleteProps.label = "test"
    render(<Autocomplete {...mockAutocompleteProps} />);
    const wrapper = await screen.findByTestId('wrapper');
    await userEvent.click(wrapper)
    const listbox = await screen.findByRole('list', {
      name: /datalist/i,
    })

    expect(listbox).toBeTruthy();

    const items = await within(listbox).findAllByRole("listitem")
    expect(items.length).toBe(4)

    const itemNames = items.map(item => item.textContent)
    expect(itemNames).toMatchSnapshot(`
      Array [
        'black', 'white', 'red', 'yellow'
      ]
    `)
  });
});
