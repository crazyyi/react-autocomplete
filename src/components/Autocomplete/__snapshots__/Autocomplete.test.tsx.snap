// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autocomplete component renders list items using renderItem prop: 
      Array [
        'black', 'white', 'red', 'yellow'
      ]
     1`] = `
Array [
  "",
  "",
  "",
  "",
]
`;

exports[`Autocomplete component should match snapshot 1`] = `
<div>
  <div
    class="autocomplete-wrap"
    data-testid="wrapper"
  >
    <input
      data-testid="input"
    />
    <div
      data-testid="error"
    />
  </div>
</div>
`;
